### DISCUSSION TITLE
incorrect styles across component boundaries

### DISCUSSION THREAD

**Discussion by jjenzz** [css-hooks/css-hooks#43] [General]
i noticed that when there are conflicting longhand/shortand properties across component boundaries, it will [sometimes result in incorrect styles](https://codesandbox.io/p/devbox/charming-https-99txml?file=%2Fsrc%2FApp.tsx). i was expecting `backgroundColor: black` to win here.

i haven't looked into the css-hooks code but on the surface it looks like it could be relying on object merging to override styles which can cause this issue:

![Screenshot 2023-12-01 at 00 24 01](https://github.com/css-hooks/css-hooks/assets/175330/822fcd6d-687d-4104-8bea-ff09df0c9682)

note how `background: 'orange'` comes **after** `backgroundColor: 'black'` in the resulting object.

love what you're doing over here btw 👍 i have been exploring similar ideas with a css lib of my own and it's very reassuring to find others exploring `style` attribute solutions ☺️

**Comment by nsaunders** [comment]
Hi Jenna, thanks for raising this issue.

You are right to be concerned about mixing shorthand and longhand properties. CSS Hooks makes no effort to expand shorthand properties, which can lead to some surprising results. I've had it in the back of my mind that there should probably be something like an ESLint plugin to help people avoid this pitfall.

With that said, if I'm understanding it correctly, the issue isn't caused by or limited to CSS Hooks. I've removed the `hooks()` calls in your example and am still seeing the unexpected result: [updated example](https://codesandbox.io/p/devbox/naughty-tristan-xck349?file=%2Fsrc%2FApp.tsx%3A12%2C1)

If I've missed something, please let me know!

**Reply by nsaunders** [reply]
Following up on my previous example, I've put together a [side-by-side comparison](https://codesandbox.io/p/devbox/trusting-roman-n6mdp7?file=%2Fsrc%2FApp.tsx%3A11%2C19) of what happens when using object-spread syntax to override styles vs. a specialized `merge` function that filters out the "base" declarations that should be overridden. So it looks like @b3nten was on the right track, at least with regard to the original issue @jjenzz pointed out.

**Comment by robinweser** [comment]
I'd say this is an issue not only with inline style but with CSS in general. I once wrote about it: https://weser.io/blog/the-shorthand-longhand-problem-in-atomic-css
There's https://github.com/robinweser/inline-style-expand-shorthand that helps with some of it, but I wouldn't recommend using it necessarily. Personally, I'd rather enforce longhands wining in whatever way. For CSS there're repeated selectors, for inline styles I assume we can just sort the object?

Check [this example](https://codesandbox.io/p/devbox/wandering-glitter-6srfdn?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpy8tvis00063d5w6yeuwi8m%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpy8tvir00033d5wgw0g2fsp%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpy8tvir00043d5wovrk7tuu%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpy8tvis00053d5wcnvhqt7j%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B60%252C40%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpy8tvir00033d5wgw0g2fsp%2522%253A%257B%2522id%2522%253A%2522clpy8tvir00033d5wgw0g2fsp%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clpy8tvis00053d5wcnvhqt7j%2522%253A%257B%2522id%2522%253A%2522clpy8tvis00053d5wcnvhqt7j%2522%252C%2522activeTabId%2522%253A%2522clpy8usym00ap3d5w0m3axekm%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522id%2522%253A%2522clpy8usym00ap3d5w0m3axekm%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clpy8tvir00043d5wovrk7tuu%2522%253A%257B%2522id%2522%253A%2522clpy8tvir00043d5wovrk7tuu%2522%252C%2522activeTabId%2522%253A%2522clpy8uny6008v3d5wxij57wzy%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clpy8uny6008v3d5wxij57wzy%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) with some very rudimentary sorting applied.

**Comment by jjenzz** [comment]
> the issue isn't caused by or limited to CSS Hooks

correct, this is a side-effect of `Object.assign` however, when people use styling libraries for composition across component boundaries the expectation is usually that "the last style applied always wins".

**Comment by b3nten** [comment]
Can this be fixed with a custom merge function?

**Comment by jesstelford** [comment]
@robinweser Sorting is a good idea. I tried alphabetical to start, then realised there are some special cases that it doesn't work for. Specifically `flexFlow` which is a shorthand for `flexDirection flexWrap` gets sorted to `flexDirection flexFlow flexWrap` :(

Possibly defining a custom sort order would work, possibly by sorting alphabetically first, then re-sorting the special cases?

**Reply by robinweser** [reply]
Good catch! Was more to proof that it works in general rather than a proper solution. I‘m working on a toolbelt for css-hooks that solved this properly using something similar to https://github.com/robinweser/fela/tree/master/packages/fela-enforce-longhands

**Reply by b3nten** [reply]
my gut feeling is that we should always respect the order provided by the user.

**Reply by nsaunders** [reply]
> my gut feeling is that we should always respect the order provided by the user.

To me, this is the right default, but I could imagine opt-in sorting being useful. Otherwise, I was thinking maybe a linter could help with shorthand issues?

**Reply by robinweser** [reply]
I'm not sure about the right defaults to be honest. Consider the following style:
```js
const style = {
 backgroundColor: 'blue',
  background: 'green',
  '@media (min-width: 800px)': {
    backgroundColor: 'yellow',
  },
}
```

What background color would you expect here?
I guess we can agree that `green` overwrites `blue` as per CSS standard.
But what happens at width >= 800px? I would expect it to turn `yellow`, but with css-hooks it will stay `green` no matter what as media queries are all resolved into the same property eventually. So, either you have to stay consistent with your shorthand-longhand usage which is arguably very hard if you allow style extension (think of a component library for example).
or you enforce a consistent and deterministic resolution pattern.

(Or we finally remove shorthand properties from the CSS specs 🤓)

**Comment by jjenzz** [comment]
this could maybe be solved with something like this as long as the `css` utility is always used for styling: https://codesandbox.io/p/sandbox/hopeful-wave-kyhyzw?file=%2Fsrc%2FApp.js%3A17%2C19

obviously this doesn't go as far as handling nested styles but hopefully the gist is there for you to expand upon.

**Reply by nsaunders** [reply]
Thanks @jjenzz, I've been resisting the idea of a merge function for fear of "scope creep", but I think you've demonstrated that it can be done inexpensively and would go a long way in helping people avoid a common pitfall.

As for merging nested styles, I have a quick question for you. Given a "base" style `{ color: "blue", "&:hover": { color: "lightblue" } }` and an "override" style `{ color: "black" }`, what would you expect the result to look like?

**Reply by jjenzz** [reply]
i'd expect it to be `black` and turn `lightblue` on hover. essentially, adding an override would behave the same as it would if we added it as a new style rule to the bottom of a stylesheet:  https://codepen.io/jjenzz/pen/qBvZXeb

**Comment by nsaunders** [comment]
I believe the new sorting options in v2 address this issue.
